stellargraph/core/graph_standard.py

```diff
@@ -0,0 +1,213 @@
+    """
+
+    def __init__(self, is_directed=False, edges=None, nodes=None, **kwargs):
+        self._is_directed = is_directed
+        self._node_data = nd = self._get_node_data(nodes, kwargs)
+        self._edge_data = ed = self._get_edge_data(edges, kwargs)
+        self._edge_cache = EdgeCache(kwargs.get("edge_cache"), ed, nd)
+
+    @staticmethod
+    def _get_node_data(data, kwargs):
+        attrs = {}
+        for name in ["node_id", "node_type", "node_features", "default_node_type"]:
+            value = kwargs.get(name)
+            if value is not None:
+                attrs[name] = value
+        return NodeData(data, **attrs)
+
+    @staticmethod
+    def _get_edge_data(data, kwargs):
+        attrs = {}
+        for name in [
+            "source_id",
+            "target_id",
+            "edge_id",
+            "edge_type",
+            "edge_weight",
+            "default_edge_type",
+            "default_edge_weight",
+        ]:
+            value = kwargs.get(name)
+            if value is not None:
+                attrs[name] = value
+        return EdgeData(data, **attrs)
+
+    ##################################################################
+    # Common interface:
+
+    def is_directed(self) -> bool:
+        return self._is_directed
+
+    def number_of_nodes(self) -> int:
+        return self._node_data.num_nodes()
+
+    def number_of_edges(self) -> int:
+        return self._edge_data.num_edges()
+
+    def nodes(
+        self, include_info: bool = False
+    ) -> Union[Iterable[Any], Iterable[tuple]]:
+        return self._node_data.nodes(include_info)
+
+    def edges(self, include_info: bool = False) -> Iterable[Any]:
+        return self._edge_data.edges(include_info)
+
+    def has_node(self, node: Any) -> bool:
+        return self._node_data.has_node(node)
+
+    def neighbour_nodes(self, node: Any) -> Iterable[Any]:
+        return set(self._edge_cache.neighbour_nodes(node))
+
+    def in_nodes(self, node: Any) -> Iterable[Any]:
+        if self.is_directed():
+            return set(self._edge_cache.in_nodes(node))
+        return self.neighbour_nodes(node)
+
+    def out_nodes(self, node: Any) -> Iterable[Any]:
+        if self.is_directed():
+            return set(self._edge_cache.out_nodes(node))
+        return self.neighbour_nodes(node)
+
+    def node_type(self, node: Any) -> Any:
+        return self._node_data.node_type(node)
+
+    ##################################################################
+    # Computationally intensive methods:
+
+    def info(self, show_attributes: bool = True, sample: Optional[int] = None):
+        directed_str = "Directed" if self.is_directed() else "Undirected"
+        s = "{}: {} multigraph\n".format(type(self).__name__, directed_str)
+        s += " Nodes: {}, Edges: {}\n".format(
+            self.number_of_nodes(), self.number_of_edges()
+        )
+        # TODO Mimic NetworkXStellrGraph output
```

Comment:
BTW I have issues with the current info method as it parses the graph to identify the schema. I think in the future we should just report graph info that doesn't require parsing the graph, it's really annoying for the user to type `G.info()` and have to wait for 10 mins!

Also, in the future I think we should move to a pre-calculated schema in the graph class so we can display the schema info.

Reply:
Agreed, but at the time I was being careful not to just remove any functionality that I didn't understand. I vaguely think `info()` exists purely because NetworkX has one?

Reply:
And everywhere you look in the code you will find `create_graph_schema()`, so it might not be trivial to get rid of this.

Reply:
You are right, the main reason `.info` exists is it is extending the `.info` of the NetworkX graph.

No, it won't be trivial but this is an opportunity to relieve us of some technical debt so I wouldn't be constrained by us doing it another way.

Reply:
I'm certainly not arguing against doing it! Just that a lot of code will initially be broken. Some will be trivial (find and delete `info()` calls, but I think the graph schema stuff is going to be tricker.